---
title: "SOCIAL"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SOCIAL}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```
## Single-cell transcriptOmics Cell-cell Interaction ALgorithm

**Attention**

The tutorial provided uses very small datasets to demonstrate how to use SOCIAL. Due to the large memory required to analyze single-cell transcriptomics, we require that users run SOCIAL on a HPC supercomputing cluster, such as the NIH Biowulf.

## 1. Installation

```{r install}
require(devtools)
devtools::install_github("sahil-sahni/5. Package(s)/SOCIAL")
```

## 2. Library

```{r library}
library(SOCIAL)
```

## 3. Organized inputs
SOCIAL inputs require specific formatting.

### expr
expression matrix (TPM), rows are genes, columns are cells. Set rownames to gene symbols and column names to cell ids.

```{r expr}
expr = matrix(c(1:9), nrow=3, ncol=3, 
              dimnames = list(c('CXCL9','CXCL10', 'CXCR3'), 
                              c('cell1','cell2', 'cell3')))
print(expr)
```

### ct_map 
dataframe describing the mapping of each cell to individual and cell type. Note that for cohort studies user can label all the cells with their respective patient to get ligand-activity at a per patient level:

```{r ct_map1}
ct_map = data.frame(Cell=c('cell1','cell2','cell3'), 
                    cell_type=c('Mal','CD8T','NK'), 
                    samples = c('patientA','patientB','patientC')) 
```

or user can label all the samples the same to evaluate enrichment of each cell-type-specific ligand-receptor interactions across all TME in a specific biological system (make sure for step 4 to set ```median_cutoff=F```):

```{r ct_map2}
ct_map = data.frame(Cell=c('cell1','cell2','cell3'),
                    cell_type=c('Mal','CD8T','NK'), 
                    samples = c('cohortA','cohortA','cohortA')) 
```

or user can label all the samples based on their treatment response to evaluate enrichment of each cell-type-specific ligand-receptor interactions across different treatment groups (make sure for step 4 to set ```median_cutoff=F```):

```{r ct_map3}
ct_map = data.frame(Cell=c('cell1','cell2','cell3'), 
                    cell_type=c('Mal','CD8T','NK'), 
                    samples = c('R','R','NR')) 
```

users can also use downsampling to aggregate pseudopatients as in Sahni et al. (see SOCIAL_Jerby-Arnon.Rmd)

## 4. SOCIAL
### Step 1: Query the LIRICS database
Initially, we queried the LIRICS database to identify plausible ligand-receptor interactions. Output is a matrix with first column corresponding to ligand genes and second corresponding to receptor genes.

download [LIRICS.db](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8983586/bin/NIHMS1770717-supplement-2.xlsx) excel here

```{r}
## SOCIAL Step 1 ## 

# data from Kun Wang Cancer Discovery 2022
lirics.db = readxl::read_excel("NIHMS1770717-supplement-2.xlsx",
                               sheet = "ligand_receptor_interactions", 
                               skip=1) 

lirics.db = lirics.db %>% 
  dplyr::select(ligand, receptor) %>% # select ligand and receptor genes
  toupper(.) # make sure genes are uppercase!!

# subset genes within ligand-receptor complexes that are shared with LIRICS database
pairs = SOCIAL.query_LRdb(lirics.db, expr)
```

### Step 2-3: Compute interaction score and permutation test.
Next, we computed the ligand-receptor interaction score by multiplying the average expression levels of the ligand and receptor complexes for each interaction pair and cell type. Following that, we performed permutation tests (utilizing 100 iterations in our study) by randomly shuffling cell type labels. This allowed us to derive empirical p-values by calculating the fraction of permutation tests resulting in a higher interaction score than the foreground score determined in step 2. A lower p-value suggests a higher likelihood of the interaction occurring.

We require users to run SOCIAL step 2 on a HPC supercomputing cluster to harness parallelization with rslurm:

```{r}
## SOCIAL Step 2-3 with HPC clustering ## 
n_iterations = 100 # bootstrapping null distribution (100 recommended)
interaction_score = SOCIAL.cis_rslurm(expr, ct_map, pairs, n_iterations = n_iterations)
```

### Step 4: Accomadate for variations across samples

Optionally, ligand-receptor interactions can be further denoted as significantly activated if the average expression level of both the ligand and receptor genes is greater than the median across all samples.

```{r}
## SOCIAL Step 4 ## 
SOCIAL_output = SOCIAL.infer_activity(interaction_score, # SOCIAL step 2-3 output
                      p_cutoff = 0.05, # recommended cut-off
                      median_cutoff = T) # recommended if 2+ more unique samples are characterized 
SOCIAL_output
```
