---
title: "SPECIAL"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SPECIAL}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval=F
)
```
## SPatial transcriptomics cEll-Cell Interaction ALgorithm

**Attention**

The tutorial provided uses very small datasets to demonstrate how to use SPECIAL (+SOCIAL). Due to the very large memory required to analyze spatial transcriptomics, we require that users run SPECIAL on a HPC supercomputing cluster, such as the NIH Biowulf.

## 1. Installation

```{r install}
require(devtools)
devtools::install_github("sahil-sahni/5. Package(s)/SOCIAL")
```

## 2. Library

```{r library}
library(SOCIAL)
```

## 3. Organized inputs
SPECIAL inputs require specific formatting. These inputs are similar to SOCIAL, with the exception that location information is accounted for.

### expr
expression matrix (TPM), rows are genes, columns are cells. Set rownames to gene symbols and column names to cell ids.

```{r expr}
expr = matrix(c(1:9), nrow=3, ncol=3, 
              dimnames = list(c('CXCL9','CXCL10', 'CXCR3'), 
                              c('cell1','cell2', 'cell3')))
print(expr)
```

### loc
Data.frame describing the mapping of each cell to cell type and location. Direct output of ***CytoSPACE***. Please keep at least CytoSPACE output columns: OriginalCID, CellType, row, and col.

```{r ct_map1}
ct_map = data.frame(UniqueCID=c('CS1','CS2','CS3'), # unique single-cell barcode generated by CytoSPACE
                    OriginalCID=c('cell1','cell2','cell3'), # original single-cell barcode generated from study (KEEP)
                    CellType = c('NK','Bcell','Mal'), # cell type of individual single-cell (KEEP)
                    SpotID = c('spot1','spot2','spot3'), # barcode or coordinate of the spatial location of the aligned single-cell 
                    row = c(1,2,3), # y coordinate of aligned single-cell (KEEP)
                    col = c(3,2,1) # x coordinate of aligned single-cell (KEEP)
                    ) 
```

### pairs
Matrix with first column (V1) corresponding to the ligand genes and second column corresponding to the receptor genes (complex with up-to 3 genes seperated by ";", i.e. "A;B;C"). This data.frame does not have to be prefiltered, and could be a database (equivalent to db) of plausible ligand-receptor pairs (as in SOCIAL.query_LRdb). Note in Sahni et al. 2024, pairs input was prefiltered using SOCIAL.query_LRdb before running SPECIAL (this is no longer necessary), but the db must be labeled as pairs if using a SPECIAL input file.

Download [LIRICS.db](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8983586/bin/NIHMS1770717-supplement-2.xlsx) excel here.

```{r pairs}

# data from Kun Wang Cancer Discovery 2022
lirics.db = readxl::read_excel("NIHMS1770717-supplement-2.xlsx",
                               sheet = "ligand_receptor_interactions", 
                               skip=1) 

lirics.db = lirics.db %>% 
  dplyr::select(ligand, receptor) %>% # select ligand and receptor genes
  toupper(.) # make sure genes are uppercase!!

pairs = lirics.db
```

## 4. SPECIAL
### Step 1 & 2: Divide spatial slides into regions and run SOCIAL step 1 through 3 on each region 
Step 1 utilizes either a sliding window or k-means clustering approach on bulk (i.e. Visium 10X and Legacy) and SlideSeqV2 spatial transcriptomics, respectively, to divide spatial slides into “regions” of approximately 250 μm in diameter. Step 2 employs SOCIAL steps 1 through 3 to infer cell-type-specific interaction activity within each ~250 μm region.

You'll also need specify arguments regarding the spatial dimension of the data including:

1. **```distance```**: distance (in unit "x") of individual spatial regions. For bulk platforms (i.e. Visium or Legacy) it is the radius of the regions. For SlideSeqV2 it is the diameter of the regions.
2. **```platform```**: whether it is "slideseqv2" or "bulk" (i.e. Visium or Legacy ST)
3. **```puck_diameter```**: diameter (in unit "x") of the SlideSeqV2 puck. For SlideSeqV2 ONLY!

We require users to run SPECIAL step 1 & 2 on a HPC supercomputing cluster to harness parallelization with rslurm:

```{r}
## SPECIAL Step 1-2 with HPC clustering ## 

n_iterations = 100 # bootstrapping null distribution (100 recommended)
name = 'tumorslide1' # name of tumor slide
distance = 1
platform = 'bUlK' # not case sensitive

SPECIAL_output = SPECIAL(expr, 
                         pairs, # doesn't need to be prefiltered
                         loc,
                         name, 
                         pairs, 
                         n_iterations,
                         distance,
                         platform)
```

### Step 4: Accomadate for variations across regions

Ligand-receptor interactions are further denoted as significantly activated if the average expression level of both the ligand and receptor genes is greater than the median across all regions

```{r}
## SPECIAL Step 4 ## 
SPECIAL_activity = SOCIAL.infer_activity(SPECIAL_output, # SPECIAL output
                      p_cutoff = 0.05, # recommended cut-off
                      median_cutoff = T) # recommended if 2+ more unique samples are characterized 
SPECIAL_activity
```
